Update service broadcast settings

idavidmcdonald opened the pull request:
https://www.pivotaltracker.com/story/show/176459886

This route can be used by the admin interface to change all the settings for a broadcast service.

Admin PR to follow

Best reviewed commit by commit

klssmith commented on app/service/service_broadcast_settings_schema.py:
```diff
@@ -0,0 +1,12 @@
+    "properties": {
+        "broadcast_channel": {"enum": ["test", "severe"]},
+        "service_mode": {"enum": ["training", "live"]},
+        "provider_restriction": {"enum": [None, "three", "o2", "vodafone", "ee"]}
```
Should it be `null` instead of `None`? I haven't got the admin part of this to try it out, so wasn't sure if it needs to be something like `"type": ["string", "null"], "enum": [ null, "three", "o2", "vodafone",  "ee"]`

idavidmcdonald replied:
Good question and I'm not 100% sure. Given that my local dev works fine when the admin app posts to it, I think it's OK.

I think what is happening is that when I send `None`, when it's parsed into JSON, it becomes `null`. And then when it's parsed back to python it becomes `None` and that appears to be validated correctly against the list of enum options.

If I change it to `"provider_restriction": {"type": ["null"]}` and then run `test_set_as_broadcast_service_updates_mobile_provider_restriction` it passes for the `None` case. 

Therefore I think if I changed it to `"type": ["string", "null"], "enum": [ None, "three", "o2", "vodafone", "ee"]` then it would be equivalent to what I already have written.

Note, `"type": ["string", "null"], "enum": [ null, "three", "o2", "vodafone", "ee"]` would be invalid because `null` is not defined and this is python code being evaluated.

idavidmcdonald replied:
Oh, the other thing I could have done is to remove `provider_restriction` from the list of required keys but the downside of this is that then it makes it a bit opaque that if you didn't include this key, it might remove any current setting in the database for the provider restriction for this service. I think therefore it's good to make the caller of the API be very explicit about what they want to change by including the `provider_restriction` key

klssmith commented on tests/app/service/test_rest.py:
```diff
@@ -3645,3 +3649,365 @@ def test_get_returned_letter(admin_request, sample_letter_template):
+    data = {
+        'broadcast_channel': channel,
+        'service_mode': 'live',
+        'provider_restriction': None,
```
I haven't got the admin part of this to check, but could the data ever come through from admin as `"None"` (a string)?

idavidmcdonald replied:
So I've got it up and running locally with the admin app doing

```
data = {
            "service_mode": service_mode,
            "broadcast_channel": broadcast_channel,
            "provider_restriction": None
        }

        return self.post("/service/{}/set-as-broadcast-service".format(service_id), data)
```

And the API handles it well.

What I will do though is add a test to make sure that the string `"None"` is rejected by the API

idavidmcdonald replied:
Have pushed up a rebased commit with the `"None"` test case added

idavidmcdonald commented on tests/app/service/test_rest.py:
```diff
@@ -3645,3 +3649,365 @@ def test_get_returned_letter(admin_request, sample_letter_template):
+):
+    current_permissions = [p.permission for p in sample_service.permissions]
+    assert len(current_permissions) > 0
+    assert current_permissions != [BROADCAST_TYPE]
```
Good idea, that's a stronger assertion. Have added

idavidmcdonald commented on tests/app/service/test_rest.py:
```diff
@@ -3645,3 +3649,365 @@ def test_get_returned_letter(admin_request, sample_letter_template):
+
+
+def test_set_as_broadcast_service_maintains_broadcast_permission_for_existing_broadcast_service(
+    client, notify_db, sample_broadcast_service, broadcast_organisation
```
Good spot, thank you

idavidmcdonald commented on tests/app/service/test_rest.py:
```diff
@@ -3645,3 +3649,365 @@ def test_get_returned_letter(admin_request, sample_letter_template):
+        'provider_restriction': None,
+    }
+
+    resp = client.post(
```
Good idea, this was worth doing

idavidmcdonald commented on app/dao/service_broadcast_settings_dao.py:
```diff
@@ -0,0 +1,15 @@
+from app.models import ServiceBroadcastSettings
+
+
+def insert_or_update_service_broadcast_settings(service, channel, provider_restriction=None):
```
So this is interesting...

Why do we want to commit the transaction immediately after `insert_or_update_service_broadcast_settings` is run? As opposed to waiting for the end of the `set_as_broadcast_service` where everything outstanding gets committed.

I can see that
`dao_remove_service_permission` is not transactional
`dao_add_service_permission` is transactional
`dao_update_service` is transactional
`dao_add_service_to_organisation` is transactional

What makes us decide if a function should be transactional?

I can see some cases for wanting a function to be transactional. When we receive a post request to send an email, we want to save the email to the DB and then put a task on a queue to send the email. It makes sense the have the saving of the email to the DB as transactional because if we put the task on the queue and then attempt to commit the transaction after this and it fails, we are in a dodgy state where we've put a task on a queue for something that doesn't exist.

I don't see a similar case here... but maybe there is a reason still that we want it to be transactional? Maybe it's even just for consistency?

As an interesting aside, I was thinking about the fact that if `dao_remove_service_permission` became transactional then my code would remove all permissions and then would add the permission back for broadcasts. This means if someone already has the broadcast permission it will remove it and then add it again and so you have a few milliseconds without the permission (and whether that would be OK)

klssmith replied:
`dao_remove_service_permission` doesn't have the transactional decorator but it still [commits the session in the function](https://github.com/alphagov/notifications-api/blob/9b940637a536d1716411209674122dce1a54c652/app/dao/service_permissions_dao.py#L21), which does the same thing. I thought we generally followed the pattern of always committing after DAO functions that add or modify data, but maybe there are a few exceptions where we don't want to do that.

If the function is going to be kept without the transactional decorator, it might be worthwhile adding a comment to the function pointing that out (since it's not how many of the other DAO functions work, so it might be easy to miss that there is no database `commit` happening)

idavidmcdonald replied:
Ah, I missed the inline commiting on `dao_remove_service_permission`.

That all seems very consistent then and therefore seems like a good case for me to add the transactional decorator. Will do this now

idavidmcdonald replied:
Also given this, any thought on the fact that for a few milliseconds, the broadcast permission will have been removed completely, if they already had it, before being added again? I could change my code to not remove the broadcast permission for example. I don't think it's a problem to be honest, the risk of that causing a problem is very very small but thought I'd get another opinion please

idavidmcdonald replied:
I've added the transactional decorator and force pushed

servingUpAces replied:
The question to ask when wrapping db operations with @transactional is would the data be in the right state if the first statement committed but the second statement failed. Would the service make sense if it had the broadcast settings but not the permission?

servingUpAces replied:
I think in this case all the updates/inserts should be wrapped in a single transaction rather than several separate ones.

idavidmcdonald replied:
Thank you Becca, I agree. Will make this change

idavidmcdonald replied:
Change pushed up in new commit

idavidmcdonald commented:
@klssmith, questions/proposal I like to get your opinion on please

- What should we do about setting the 'go_live_at': and 'go_live_user' fields?
- Should we set the `go_live_user` field at all as that is usually used for the person who requested the go live I believe which we won't have in this case (they won't submit a form to request to go live, we will instead create them a new service which is live for them to use once requested by the Cabinet Office team)
- Should we update the `go_live_at` date if their account type was changed from one type of live service to another?
- Should we remove the `go_live_at` date if their account type was changed from live to training mode?
- If we don't set either of the two fields, maybe we should have an audit event instead?

My current proposal is that
- we set the `go_live_at` date when the service is made live and remove it if the service is put back into training mode
- we set the `go_live_user` to be the user who made the change in the admin app and remove it when they get put back to training mode
- we don't add an audit event for the moment as we can probably get away without out it for a few weeks but we likely do as part of this story https://www.pivotaltracker.com/story/show/176698377

idavidmcdonald commented:
Right @klssmith. Thanks for your comments and questions. I think this is ready for a second review please.

I've 
- made a few changes based on your comments
- added one new commit 'Hardcode service broadcast channel that API shows' to fix a bug/edge case in our code
- added one new commit 'Add `go_live_at` timestamp to set_as_broadcast_service' which sets the go_live_at timestamp. You'll see in this one that I decided not to also record the `go_live_user` as I don't think it's worth it and will be replaced by the more important audit events in the future. Feel free to disagree on this though!

If you want to have a play around locally, I've pushed up a very WIP branch `broadcast-service-settings` on the admin app that will interact with this endpoint.

servingUpAces commented on app/dao/broadcast_service_dao.py:
```diff
@@ -0,0 +1,56 @@
+
+from app import db
+from app.models import ServiceBroadcastSettings, ServicePermission, Organisation, BROADCAST_TYPE
+from app.dao.service_permissions_dao import dao_fetch_service_permissions
```
Some unused import here.

servingUpAces commented on app/dao/broadcast_service_dao.py:
```diff
@@ -0,0 +1,56 @@
+        service.go_live_at = None
+
+    service.restricted = True
+    if service_mode == "live":
```
why isn't this part of the previous `if service_mode == "live"`?

servingUpAces commented on app/dao/service_broadcast_settings_dao.py:
```diff
@@ -0,0 +1,15 @@
+from app.models import ServiceBroadcastSettings
+
+
+def insert_or_update_service_broadcast_settings(service, channel, provider_restriction=None):
```
Does this method need to be in a different file? 
Can service_broadcast_settings_dao and broadcast_service_dao be merged into one file?

servingUpAces commented on app/service/rest.py:
```diff
@@ -29,7 +29,10 @@
-from app.dao.organisation_dao import dao_get_organisation_by_service_id
+from app.dao.organisation_dao import (
+    dao_get_organisation_by_service_id,
+    dao_add_service_to_organisation,
```
unused import

servingUpAces commented on app/service/rest.py:
```diff
@@ -42,6 +45,12 @@
     dao_get_contact_list_by_id,
     save_service_contact_list,
 )
+from app.dao.service_permissions_dao import (
```
unused import

servingUpAces commented on app/service/rest.py:
```diff
@@ -73,6 +82,9 @@
     dao_update_service,
     get_services_by_partial_name,
 )
+from app.dao.service_broadcast_settings_dao import (
```
unused import
